packages/controllers/src/permissions/Permission.ts

```diff
@@ -0,0 +1,443 @@
+   * one context in our usage (i.e. the user's MetaMask instance).
+   */
+  readonly '@context'?: NonEmptyArray<string>;
+
+  // TODO:TS4.4 Make optional
+  /**
+   * The caveats of the permission.
+   * @see {@link Caveat} For more information.
+   */
+  readonly caveats: null | NonEmptyArray<CaveatConstraint>;
+
+  /**
+   * The creation date of the permission, in UNIX epoch time.
+   */
+  readonly date: number;
+
+  /**
+   * The GUID of the permission object.
+   */
+  readonly id: string;
+
+  /**
+   * The origin string of the subject that has the permission.
+   */
+  readonly invoker: OriginString;
+
+  /**
+   * A pointer to the resource that possession of the capability grants
+   * access to.
+   *
+   * In the context of MetaMask, this is always the name of an RPC method.
+   */
+  readonly parentCapability: string;
+};
+
+/**
+ * A `ZCAP-LD`-like permission object. A permission is associated with a
+ * particular `invoker`, which is the holder of the permission. Possessing the
+ * permission grants access to a particular restricted resource, identified by
+ * the `parentCapability`. The use of the restricted resource may be further
+ * restricted by any `caveats` associated with the permission.
+ *
+ * See the documentation for details.
+ *
+ * @template TargetKey - They key of the permission target that the permission
+ * corresponds to.
+ * @template AllowedCaveat - A union of the allowed {@link Caveat} types
+ * for the permission.
+ */
+export type ValidPermission<
+  TargetKey extends TargetName,
+  AllowedCaveat extends CaveatConstraint,
+> = ExtractPermissionTargetNames<TargetKey> extends never
+  ? never
+  : PermissionConstraint & {
+      // TODO:TS4.4 Make optional
+      /**
+       * The caveats of the permission.
+       * @see {@link Caveat} For more information.
+       */
+      readonly caveats: AllowedCaveat extends never
+        ? null
+        : NonEmptyArray<AllowedCaveat> | null;
+
+      /**
+       * A pointer to the resource that possession of the capability grants
+       * access to.
+       *
+       * In the context of MetaMask, this is always the name of an RPC method.
+       */
+      readonly parentCapability: ExtractPermissionTargetNames<TargetKey>;
+    };
+
+/**
+ * A utility type for ensuring that the given permission target name conforms to
+ * our naming conventions.
+ *
+ * See the documentation for the distinction between target names and keys.
+ */
+type ValidTargetName<Name extends string> = Name extends `${string}*`
+  ? never
+  : Name extends `${string}_`
+  ? never
+  : Name;
+
+/**
+ * A utility type for extracting permission target names from a union of target
+ * keys.
+ *
+ * See the documentation for the distinction between target names and keys.
+ *
+ * @template Key - The target key type to extract target names from.
+ */
+export type ExtractPermissionTargetNames<Key extends string> = ValidTargetName<
+  Key extends `${infer Base}_*` ? `${Base}_${string}` : Key
+>;
```

Comment:
@Gudahtt 

> It does not seem possible to me for `ExtractPermissionTargetNames<TargetKey>` to evaulate to never though. It literally doesn't return never, making this condition wholly pointless.

I added the `ValidTargetName` wrapping inside of `ExtractPermissionTargetNames`, so `ExtractPermissionTargetNames<TargetKey>` now can (and does) evaluate to `never` if the input is invalid.

Reply:
Right, I see. I still think it has no business being in the `ValidPermission` type as a ternary. It makes the type more complicated for no reason. If this is broken and evaluates to `never`, that broken-ness will be expressed in the type via the `parentCapability` field.

Reply:
Alright, removed that ternary here: 48a2516